lean4/Synthesis/Utils.lean

```diff
@@ -49,18 +49,18 @@ partial def readString (stream : IO.FS.Stream) (buf : ByteArray) : IO ByteArray
     readString stream buf
 
 /-- Reads from the stream and returns a list of lines. -/
-partial def readFile (stream : IO.FS.Stream) (buf : List String) : IO (List String) := do
-  let line ← stream.getLine
-  if line.isEmpty then
-    pure buf
-  else
-    let mut buf := buf.append [line]
-    readFile stream buf
+partial def readFile (stream : IO.FS.Stream) : IO (List String) := do
+  let mut ret : Array String := #[]
+  let mut line : String ← stream.getLine
+  while ! line.isEmpty do
+    ret := ret.push line
+    line ← stream.getLine
+  return ret.toList
 
 /-- Reads from the list of streams and returns a list of all lines. -/
 partial def readAll (streams : List IO.FS.Stream) (buf : List String) : IO (List String) := do
   match streams with
   | [] => pure buf
-  | stream :: rest => 
-    let buf ← readFile stream buf
-    readAll rest buf
+  | stream :: rest =>
+    let buf' ← readFile stream
+    readAll rest (buf ++ buf')
```

Comment:
This is a candidate for further optimization: `readFile` should return an `Array` and this should use `Array.append`.

Reply:
Changed it all to `Array`s in commit 5bbb02, and I'm using `.toListImpl` to convert it back to a list in some of the aggregators for now.